### PR TITLE
auth/exec: prevent multiple calls when expirationTimestamp is invalid

### PR DESCRIPTION
**Pull request by enj**
This change addresses an edge case in our connection closing logic.

Consider the following scenario:

1. Exec plugin is configured
2. Exec plugin uses certificate based auth
3. Exec plugin returns a new certificate per invocation
4. Exec plugin sets expirationTimestamp based on some external source
5. The external source and the client (i.e. kubectl) disagree on
   what the current time is
6. The external source gives the exec plugin an expirationTimestamp
   that is considered to be in the past from the perspective of the
   client (i.e. the certificate is considered to already be expired)

Call stack for a client like kubectl:

1. exec.Authenticator.UpdateTransportConfig sets up Wrap and GetCert
2. exec.roundTripper.RoundTrip is called
3. getCreds, no creds, refreshCredsLocked -> gets first certificate
4. r.base.RoundTrip is called
5. a.cert gets called to fill the TLS config with a cert
6. getCreds, "expired" creds, refreshCredsLocked -> gets second
   certificate.  sees old creds with cert and closes connections.
   This happens to be the _current_ connection and thus auth fails
   with a "use of closed connection" error.

Signed-off-by: Monis Khan <mok@vmware.com>

/kind bug
/assign @liggitt 

```release-note
Fix "use of closed network connection" error seen under particular configurations of client-go credential plugins.
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 1]
```diff
@@ -507,6 +507,17 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
...
+
+		// this method will be invoked twice for the same request if we consider the
+		// expirationTimestamp to be in the past.  if this happens when cert based
+		// auth is being used and the exec plugin returns a new cert for each invocation,
+		// we will accidentally close the in-flight connection because we think that the
+		// cert rotated (even though we never used the first cert at all).
+		if a.credsExpiredLocked() {
+			klog.InfoS("certificate was marked expired when issued, ignoring expiration until unauthorized",
+				"expirationTimestamp", a.exp.String())
+			a.exp = time.Time{}
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:519]: while this may happen to drive the behavior you want, it's pretty hard to reason about and doesn't seem correct to me

* it only gets set for cert-based auth
* it resets a.exp which is set above (which is hard to follow)
* it means credsExpired() will now return false, even though the time returned by the plugin has passed

are you wanting to make sure a credential returned from a plugin with an expirationTimestamp in the past gets used *at least once*, or that returning an expiration timestamp in the past is the same as not return an expiration timestamp at all?

Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:519]: Stepping back a bit, I think it may be simpler to frame this as:

For a particular round trip, the exec plugin must only be invoked once and only at the start of the round trip (before the base round tripper is called).

This would allow an exec plugin to return an expiration timestamp in the past, and all that would mean is that the exec plugin would be invoked on every single round trip (i.e. no caching).

This would work for both forms of auth without resetting a.exp and thus credsExpired would behave as before.

Since the exec plugin would only be called once at the start, it would be impossible for it to close the in-flight connection, because that connection does not start until the base round tripper is called.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 2]
```diff
@@ -353,40 +363,62 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
+	creds, connClosed, err := a.getCreds()
 	if err != nil {
 		return nil, err
 	}
+	if connClosed {
+		// If we close the underlying TCP connection in the process of starting the TLS handshake,
+		// there is no point in letting the handshake continue.  At least this way we provide a
+		// better error message then "write tcp host:port->host:port use of closed network connection"
+		return nil, errors.New("failing TLS handshake due to certificate rotation")
+	}
```
Comment (enj) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:404]: I am unsure of a good way to test this path, or if there is some way to avoid this altogether.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 3]
```diff
@@ -353,40 +363,62 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
+func (a *Authenticator) credsExpiredLocked() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
-func (a *Authenticator) cert() (*tls.Certificate, error) {
-	creds, err := a.getCreds()
+func (a *Authenticator) cert(ctx context.Context) (*tls.Certificate, error) {
```
Comment (enj) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:387]: If the change to this method is reverted, the new `TestTLSCredentialsCallsAndRotation` test reliably fails:

```
=== RUN   TestTLSCredentialsCallsAndRotation
=== CONT  TestTLSCredentialsCallsAndRotation
    exec_test.go:1566: unexpected exec call count: want=2, got=4
--- FAIL: TestTLSCredentialsCallsAndRotation (0.05s)
=== RUN   TestTLSCredentialsCallsAndRotation/valid_TLS_cert
I1203 13:37:56.814419   35159 exec.go:506] "credential was marked expired when issued" expirationTimestamp="2021-12-03 10:37:56 -0500 EST"
I1203 13:37:56.828111   35159 exec.go:506] "credential was marked expired when issued" expirationTimestamp="2021-12-03 10:37:56 -0500 EST"
2021/12/03 13:37:56 http: TLS handshake error from 127.0.0.1:49292: EOF
    exec_test.go:1550: got client.Get error: "write tcp 127.0.0.1:49292->127.0.0.1:49291: use of closed network connection", want nil
    --- FAIL: TestTLSCredentialsCallsAndRotation/valid_TLS_cert (0.02s)

=== RUN   TestTLSCredentialsCallsAndRotation/valid_TLS_cert_again
I1203 13:37:56.838715   35159 exec.go:506] "credential was marked expired when issued" expirationTimestamp="2021-12-03 10:37:56 -0500 EST"
I1203 13:37:56.852494   35159 exec.go:506] "credential was marked expired when issued" expirationTimestamp="2021-12-03 10:37:56 -0500 EST"
2021/12/03 13:37:56 http: TLS handshake error from 127.0.0.1:49293: EOF
    exec_test.go:1550: got client.Get error: "write tcp 127.0.0.1:49293->127.0.0.1:49291: use of closed network connection", want nil
    --- FAIL: TestTLSCredentialsCallsAndRotation/valid_TLS_cert_again (0.02s)
```

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go]: I took a shot at implementing the behavior I described above.  I am not happy with the change per se, but it seems better than what we have today.  Added tests that exercise the edge case.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go (hunk 2 above)
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:404]: I could also just turn this into a `klog` statement instead of a direct error.  That way if this logic is triggered in a case that somehow does not result in a `use of closed network connection` error, the TLS handshake would be allowed to continue.

Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:404]: I managed to expand `TestTLSCredentialsCallsAndRotation` with some keep-alive shenanigans to test this case.

**Comment by dims** [comment]
@liggitt this looks ready now.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 4]
```diff
@@ -335,13 +336,22 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 		return r.base.RoundTrip(req)
 	}
 
-	creds, err := r.a.getCreds()
+	creds, _, err := r.a.getCreds()
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:339]: Opened #108849 to track that work.

File: staging/src/k8s.io/client-go/transport/transport.go [hunk 5]
```diff
@@ -105,7 +105,7 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
 	}
 
 	if c.HasCertAuth() || c.HasCertCallback() {
-		tlsConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		tlsConfig.GetClientCertificate = func(cri *tls.CertificateRequestInfo) (*tls.Certificate, error) {
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/transport/transport.go:108]: [Yes](https://github.com/golang/go/blob/3f8a6941553ae007638335cb6af0b6a702adf64f/src/net/http/transport.go#L1334-L1342):

Screen Shot 2022-03-21 at 11 35 45 AM

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 6]
```diff
@@ -335,13 +336,22 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
 	if err != nil {
 		return nil, fmt.Errorf("getting credentials: %v", err)
 	}
 	if creds.token != "" {
+		req = utilnet.CloneRequest(req)
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:344]: > did you just notice this or did a test catch it?

I just noticed it as I was changing the code.  No test failed.

> would we want to hoist this bugfix back independent of the rest of this PR?

It would not hurt to backport though I am unsure if it matters in practice.

**Comment by enj** [comment]
@liggitt addressed all comments and rebased on master to handle conflicts.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 7]
```diff
@@ -333,54 +334,86 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
+func (a *Authenticator) credsExpiredLocked() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
-func (a *Authenticator) cert() (*tls.Certificate, error) {
-	creds, err := a.getCreds()
+func (a *Authenticator) cert(ctx context.Context) (*tls.Certificate, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:374]: since this is now returning errors in scenarios that only make sense when used as a `GetCert` callback, document this function is only used for that purpose (and maybe rename to `getCertCallback` or something to make it clear to callers?)

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 8]
```diff
@@ -333,54 +334,86 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
+	// fallback to fetching creds directly if we are not called via RoundTrip
+	creds, connClosed, err := a.getCreds()
 	if err != nil {
 		return nil, err
 	}
+	if connClosed {
+		// If we close the underlying TCP connection in the process of starting the TLS handshake,
+		// there is no point in letting the handshake continue.  At least this way we provide a
+		// better error message then "write tcp host:port->host:port use of closed network connection"
+		return nil, errors.New("failing TLS handshake due to certificate rotation")
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go]: @liggitt see if you think 13a4745349d002f0b725545f3c9bafed443babd2 is a hack we would want to live with.

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go]: (We will likely need something more sophisticated because if multiple TLS connections are established near the same time it could cause unrelated connections to be closed).
